Fix the compile error in the MinGW test

dseguin opened the pull request:
#### Summary
None

#### Purpose of change
The MinGW-Win64 test currently fails to compile the templated function in `npctalk.h`:
https://github.com/CleverRaven/Cataclysm-DDA/runs/5297671388

```
In file included from src/npctalk.cpp:56:
src/npctalk.h: In function 'std::vector<int> npcs_select_menu(const std::vector<T*>&, const string&, std::function<bool(const T*)>) [with T = npc]':
Error: src/npctalk.h:163:23: error: 'this' pointer is null [-Werror=nonnull]
  163 |                 entry += npc_list[i]->disp_name( false, true );
      |                 ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
At global scope:
cc1plus: note: unrecognized command-line option '-Wno-unknown-warning-option' may have been intended to silence earlier diagnostics
cc1plus: all warnings being treated as errors
make: *** [Makefile:948: objwin/tiles/npctalk.o] Error 1
```

#### Describe the solution
Doing what mqrause suggested below: move the definition out of the header and just use a Character pointer.

#### Describe alternatives you've considered

#### Testing
Let GHA compile and run the test.

Also tested in-game just in case:
From the `C` menu:
![c_menu](https://user-images.githubusercontent.com/12537966/155276200-fc6c3173-a170-465c-93ab-4f05f05050e9.png)
From npc dialogue:
![npc_menu](https://user-images.githubusercontent.com/12537966/155276204-3f61ee76-385a-417a-a643-e480ebcb99bb.png)


#### Additional context

mqrause commented:
I'm new to templates myself, but i think `template<class T>` is wrong and it should either be `template<class Character>` or `template<typename T>`. Ultimately it doesn't seem necessary to be a template at all and can just use Character pointers.

dseguin commented:
> I'm new to templates myself, but i think `template<class T>` is wrong and it should either be `template<class Character>` or `template<typename T>`. Ultimately it doesn't seem necessary to be a template at all and can just use Character pointers.

You're right, it doesn't make sense to have this in the header at all. Dropped the template and moved the definition out of the header.
